README
Dist::Zilla::Plugin::MetaProvides::Package

This is a "Dist::Zilla" Plugin that populates the "provides" property of
"META.json" and "META.yml" by absorbing it from your shipped modules, in
a manner similar to how "PAUSE" itself does it.

This allows you to easily create an authoritative index of what module
provides what version in advance of "PAUSE" indexing it, which "PAUSE"
in turn will take verbatim.

INSTALLATION

This is a Perl module distribution. It should be installed with whichever
tool you use to manage your installation of Perl, e.g. any of

  cpanm .
  cpan  .
  cpanp -i .

Consult http://www.cpan.org/modules/INSTALL.html for further instruction.
Should you wish to install this module manually, the procedure is

  perl Makefile.PL
  make
  make test
  make install

COPYRIGHT AND LICENSE

This software is copyright (c) 2016 by Kent Fredric
<kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Revision history for Dist-Zilla-Plugin-MetaProvides-Package

2.004004 2017-03-01T12:53:21Z 8fa7547

2.004003 2017-03-01T12:48:27Z 9169231
 [Bugfix]
 - Update to regenerate compile test which failed on perls without . in @INC
 - Closes https://github.com/kentnl/Dist-Zilla-Plugin-MetaProvides-Package/issues/9

 [Dependencies::Stats]
 - Dependencies changed since 2.004002, see misc/*.deps* for details
 - develop: +1 ↑1

2.004002 2016-08-03T16:41:52Z 37520fe
 - STABLE: This release includes changes from 2.004000 to 2.004001 and some doc fixes.
 - SHORTLIST: use Role::ModuleMetadata, include_underscore option

 [Dependencies::Stats]
 - Dependencies changed since 2.004001, see misc/*.deps* for details
 - develop: (suggests: ↑1)

 [Documentation]
 - Mistakes in synopsis about defaults clarified re: include_underscores
 - Fix confusing POD about include_underscores

2.004001 2016-07-08T03:22:43Z 3cfc89c
 - New feature: include_underscores. This is a way to reverse the change in e6106140 that removed packages with "_"
   prefixing that was requested by xdg.
 - No longer uses ConfigDumper Util.
 - No longer uses Test::KENTNL

 [Dependencies::Stats]
 - Dependencies changed since 2.004000, see misc/*.deps* for details
 - runtime: +1 -1
 - test: +2 -1 (recommends: +1)

2.004000 2016-04-27T02:37:23Z 50a9c97
 - Now using Role::ModuleMetadata to delegate actual metadata extraction from ::File objects. This should be
   effectively no change, but it could be.

 [Dependencies::Stats]
 - Dependencies changed since 2.003002, see misc/*.deps* for details
 - develop: +1
 - runtime: +1 -1

2.003002 2016-04-15T06:21:44Z 6ac5567
 - Remove QuickReference section as per #4
 - Enhance Synopsis documentation.
 - Remove unhelpful private documentation

 [Dependencies::Stats]
 - Dependencies changed since 2.003001, see misc/*.deps* for details
 - develop: +5 ↑3 -1 (suggests: ↑2)

2.003001 2015-01-15T18:39:26Z d96ad72
 [Bugfix]
 - Avoid bug in Dzil <5.009 where Role::PPI doesn't load PPI::Document

 [Dependencies::Stats]
 - Dependencies changed since 2.003000, see misc/*.deps* for details
 - configure: (recommends: ↓1)
 - develop: +2 (recommends: ↓1, suggests: ↑1)
 - runtime: +1 (recommends: +1)
 - test: (recommends: ↓2)

 [Documentation]
 - Fix false documentation about defaults (dolmen++, #5)

2.003000 2015-01-11T04:57:16Z e6ee0ca
 [Dependencies::Stats]
 - Dependencies changed since 2.002000, see misc/*.deps* for details
 - develop: +1 -1 (recommends: ↑1, suggests: ↑1)
 - runtime: +1
 - test: (recommends: ↑1)

 [Documentation]
 - Add a DESCRIPTION section for Readme purposes.
 - README is now ::Brief

 [Features]
 - Attempts to detect intentionally hidden packages and allow their presense to disable the "no packages found"
   warning. ( Closes #3 )

2.002000 2014-12-29T17:46:59Z ac1a07c
 [Dependencies::Stats]
 - Dependencies changed since 2.001003, see misc/*.deps* for details
 - develop: +1 ↑1 (recommends: ↑1, suggests: ↑1)
 - test: (recommends: ↑1)

 [Important Changes]
 - A non-Trial version mostly the same as 2.001003. I hope my de-autoboxing isn't broken

2.001003 2014-12-10T10:50:56Z d86f262
 [Dependencies::Stats]
 - Dependencies changed since 2.001002, see misc/*.deps* for details
 - configure: +1 (recommends: ↑1)
 - develop: +5 ↑1 -2 (recommends: +1, suggests: ↑2)
 - runtime: +1 ↓1 -2
 - test: +1 (recommends: ↑2)

 [Minor Internal Changes]
 - Minimum perl dropped to 5.008

 [Moderate Internal Changes]
 - Moose::Autobox nuked, and this is a trial release to make sure I didn't add any bugs.

2.001002 2014-09-04T19:04:06Z
 [00 Trivial]
 - Metadata
 - No code changes.

 [Dependencies::Stats]
 - Dependencies changed since 2.001001, see misc/*.deps* for details
 - test: (recommends: ↑1)

 [Metadata]
 - Fix copyright information

2.001001 2014-08-29T12:00:46Z
 [Bugfix]
 - avoid fail on non-build-dir

 [Dependencies::Stats]
 - Dependencies changed since 2.001000, see misc/*.deps* for details
 - test: ↑1

2.001000 2014-08-24T20:33:44Z
 [00 Minor]
 - Internals changes.
 - Tests overhaulled.
 - 5.14 package NAME VERSION BLOCK now works due to Module::Metadata 1.000022

 [Dependencies::Stats]
 - Dependencies changed since 2.000004, see misc/*.deps* for details
 - develop: +1 ↑1 (suggests: ↑2)
 - runtime: +1 ↑1
 - test: +2 ↑1

 [Internals]
 - Implement ->dump_config in terms of Util::ConfigDumper

 [Tests]
 - All tests simplified in terms of dztest and inline content.
 - 5.14 package statement tests no longer a TODO

2.000004 2014-08-15T21:17:41Z
 [00 Trivial]
 - CPANDAY!
 - no code changes.
 - tests augmented.
 - whitespace adjusted.

 [Dependencies::Stats]
 - Dependencies changed since 2.000003, see misc/*.deps* for details
 - configure: ↓1
 - develop: +1 (suggests: ↑1)
 - test: -4 (recommends: ↑1 -2)

 [Misc]
 - Improve source side POD to reduce build side whitespace.

 [Tests]
 - update Test::ReportPrereqs

2.000003 2014-07-28T23:48:39Z
 [00 Trivial]
 - Move to org
 - Update version reporter.

 [Dependencies::Stats]
 - Dependencies changed since 2.000002, see misc/*.deps* for details
 - develop: (suggests: ↑1)
 - test: +2 ↑1 (recommends: +1)

2.000002 2014-07-13T19:42:09Z
 [00 Trivial]
 - Minor packaging changes.

 [Dependencies::Stats]
 - Dependencies changed since 2.000001, see misc/*.deps* for details
 - configure: (recommends: ↑1)
 - develop: +3 -3 (suggests: +2 -1)
 - test: (recommends: ↑2)

2.000001 2014-02-06T17:13:11Z
 [00 Bugfix]
 - Need newer Dzil for encoded_content

 [Dependencies::Aggregate]
 - 5.10 : 185 → 184
 - 5.19 : 169 → 167
 - https://gist.github.com/kentfredric/8848439
 - Devel::FindPerl was dropped from Build to Test in Module::Path
 - Dist::Zilla::Util::Test::KENTNL stopped causing MB upgrade on 5.19

 [Dependencies::Stats]
 - Dependencies changed since 2.000000, see misc/*.deps* for details
 - runtime: +1

2.000000 2014-02-02T06:25:35Z
 [00 Major]
 - Toolkit moved to EUMM.
 - Tooling dependencies softened.
 - IO::String no longer used.

Build.PL

# This file was automatically generated by Dist::Zilla::Plugin::ModuleBuild v5.011.
use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.4204"
  },
  "configure_requires" => {
    "Module::Build" => "0.4204"
  },
  "dist_abstract" => "Extract namespaces/version from traditional packages for provides",
  "dist_author" => [
    "Kent Fredric <kentnl\@cpan.org>"
  ],
  "dist_name" => "Dist-Zilla-Plugin-MetaProvides-Package",
  "dist_version" => "1.15000003",
  "license" => "perl",
  "module_name" => "Dist::Zilla::Plugin::MetaProvides::Package",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Data::Dump" => "1.16",
    "Dist::Zilla::MetaProvides::ProvideRecord" => "1.14000000",
    "Dist::Zilla::Plugin::FinderCode" => 0,
    "Dist::Zilla::Plugin::MetaProvides" => "1.15000000",
    "Dist::Zilla::Role::MetaProvider::Provider" => 0,
    "IO::String" => 0,
    "Module::Metadata" => "1.000005",
    "Moose" => 0,
    "Moose::Autobox" => 0,
    "MooseX::LazyRequire" => 0,
    "MooseX::Types::Moose" => 0,
    "namespace::autoclean" => 0,
    "perl" => "5.010",
    "strict" => 0,
    "utf8" => 0,
    "warnings" => 0
  },
  "script_files" => [],
  "test_requires" => {
    "Dist::Zilla::Util::Test::KENTNL" => "0.01000011",
    "Test::Fatal" => 0,
    "Test::Moose" => 0,
    "Test::More" => "1.001002"
  }
);


my %fallback_build_requires = (
  "Dist::Zilla::Util::Test::KENTNL" => "0.01000011",
  "Module::Build" => "0.4204",
  "Test::Fatal" => 0,
  "Test::Moose" => 0,
  "Test::More" => "1.001002"
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


maint/perlcritic.rc.gen.pl
#!/usr/bin/env perl
## no critic (Modules::RequireVersionVar)

# FILENAME: bundle_to_ini.pl
# CREATED: 02/06/14 01:48:56 by Kent Fredric (kentnl) <kentfredric@gmail.com>
# ABSTRACT: Write an INI file from a bundle

use 5.008;    #utf8
use strict;
use warnings;
use utf8;

our $VERSION = 0.001;

use Carp qw( croak carp );
use Perl::Critic::ProfileCompiler::Util qw( create_bundle );
use Path::Tiny qw(path);

## no critic (ErrorHandling::RequireUseOfExceptions)
my $bundle = create_bundle('Example::Author::KENTNL');
$bundle->configure;

my @stopwords = (
  qw(
    versionless
    metadata
    conformant
    autovivification
    ),
);
for my $wordlist (@stopwords) {
  $bundle->add_or_append_policy_field( 'Documentation::PodSpelling' => ( 'stop_words' => $wordlist ) );
}
$bundle->add_or_append_policy_field(
  'Subroutines::ProhibitCallsToUndeclaredSubs' => (
    exempt_subs => join q[ ],
    qw(
      MooseX::Types::subtype
      MooseX::Types::where
      MooseX::Types::as
      MooseX::Types::ModVersion
      MooseX::Types::ProviderObject
      ),
  ),
);
$bundle->remove_policy('Documentation::RequirePodLinksIncludeText');
$bundle->remove_policy('Bangs::ProhibitDebuggingModules');
$bundle->remove_policy('CodeLayout::RequireUseUTF8');

my $inf = $bundle->actionlist->get_inflated;

my $config = $inf->apply_config;

{
  my $rcfile = path('./perlcritic.rc')->openw_utf8;
  $rcfile->print( $config->as_ini, "\n" );
  close $rcfile or croak 'Something fubared closing perlcritic.rc';
}
my $deps = $inf->own_deps;
{
  my $target = path('./misc');
  $target->mkpath if not $target->is_dir;

  my $depsfile = $target->child('perlcritic.deps')->openw_utf8;
  for my $key ( sort keys %{$deps} ) {
    $depsfile->printf( "%s~%s\n", $key, $deps->{$key} );
    *STDERR->printf( "%s => %s\n", $key, $deps->{$key} );
  }
  close $depsfile or carp 'Something fubared closing perlcritic.deps';
}



Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.008.
use strict;
use warnings;

use 5.008;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Extract namespaces/version from traditional packages for provides",
  "AUTHOR" => "Kent Fredric <kentnl\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Dist-Zilla-Plugin-MetaProvides-Package",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.008",
  "NAME" => "Dist::Zilla::Plugin::MetaProvides::Package",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Data::Dump" => "1.16",
    "Dist::Zilla" => "5.000",
    "Dist::Zilla::MetaProvides::ProvideRecord" => "1.14000000",
    "Dist::Zilla::Plugin::FinderCode" => 0,
    "Dist::Zilla::Plugin::MetaProvides" => "1.15000000",
    "Dist::Zilla::Role::MetaProvider::Provider" => 0,
    "Dist::Zilla::Role::ModuleMetadata" => 0,
    "Dist::Zilla::Role::PPI" => 0,
    "Moose" => 0,
    "MooseX::LazyRequire" => 0,
    "MooseX::Types::Moose" => 0,
    "PPI::Document" => 0,
    "Safe::Isa" => 0,
    "namespace::autoclean" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Dist::Zilla::Role::FileFinder" => 0,
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Module::Metadata" => "1.000022",
    "Path::Tiny" => 0,
    "Test::DZil" => 0,
    "Test::Fatal" => 0,
    "Test::Moose" => 0,
    "Test::More" => "0.96"
  },
  "VERSION" => "2.004004",
  "test" => {
    "TESTS" => "t/*.t t/00-compile/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "Data::Dump" => "1.16",
  "Dist::Zilla" => "5.000",
  "Dist::Zilla::MetaProvides::ProvideRecord" => "1.14000000",
  "Dist::Zilla::Plugin::FinderCode" => 0,
  "Dist::Zilla::Plugin::MetaProvides" => "1.15000000",
  "Dist::Zilla::Role::FileFinder" => 0,
  "Dist::Zilla::Role::MetaProvider::Provider" => 0,
  "Dist::Zilla::Role::ModuleMetadata" => 0,
  "Dist::Zilla::Role::PPI" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Module::Metadata" => "1.000022",
  "Moose" => 0,
  "MooseX::LazyRequire" => 0,
  "MooseX::Types::Moose" => 0,
  "PPI::Document" => 0,
  "Path::Tiny" => 0,
  "Safe::Isa" => 0,
  "Test::DZil" => 0,
  "Test::Fatal" => 0,
  "Test::Moose" => 0,
  "Test::More" => "0.96",
  "namespace::autoclean" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
